### ISSUE TITLE
MobileViT does not work with Inference with different LoRA adapters in the same batch

### ISSUE THREAD

**Issue by saeid93** [huggingface/peft#1967]
### System Info

Python 3.11.9
transformers==4.40.2
peft==0.11.2

### Who can help?

@BenjaminBossan 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder
- [ ] My own task or dataset (give details below)

### Reproduction

[MobileVit model](https://github.com/huggingface/transformers/blob/3fbaaaa64d1ef3d8327adb577994d3d11277c77a/src/transformers/models/mobilevit/modeling_mobilevit.py#L789) is not compatible with using multiple adapters in the same batch. Inferencing batches using multiple adapters using the adapter_names in the batch will trigger the following exception:

https://github.com/huggingface/peft/blob/273acf059e0f1f8bff1a3889f901475e9eb3b7ee/src/peft/tuners/lora/layer.py#L308

The root cause is that during the [unfolding operation](https://github.com/huggingface/transformers/blob/3fbaaaa64d1ef3d8327adb577994d3d11277c77a/src/transformers/models/mobilevit/modeling_mobilevit.py#L435) in the [transformers library MobileVit](https://github.com/huggingface/transformers/blob/3fbaaaa64d1ef3d8327adb577994d3d11277c77a/src/transformers/models/mobilevit/modeling_mobilevit.py#L372) the first dimension of the input is changed from `batch_size, ...` is changed to `batch_size * patch_size**2, ...` which makes it inconsistent with the `adapter_names` dimensions which is of length of `batch_size` and each entry refers to each of the batch items' adapter.

### Expected behavior

I solved this by a hack that modifies the `adapter_names` input size before sending it to the model and reverting it back to the original size for the classifier. It makes the entries proportional to the size made during the unfolding operation.

Also, we already discussed that there is a bug https://github.com/huggingface/peft/issues/1960 other than this MobileViT specific problem. Below script is the modifications needed both for https://github.com/huggingface/peft/issues/1960 and the mentioned problem together.

However, this is just a hack and I think this should work out of the box. I'm happy to investigate further when I get a chance to first solve https://github.com/huggingface/peft/issues/1960 . 

```python
# -------- changing the size of the adapter_names input ----------
        if model.base_model.model.base_model_prefix == "mobilevit":
            patch_size = model.config.patch_size
            multiply = patch_size ** 2
            resized_adapters_names = []
            for item in batch["adapter_names"]:
                multiplied = [item] * multiply
                resized_adapters_names += multiplied
            batch["adapter_names"] = resized_adapters_names
        outputs = model(**batch)

# -------- rest of the code ----------

"""
added this to solve https://github.com/huggingface/peft/issues/1960
"""

from typing import Any, Optional, Union
import torch
from torch.nn import BCEWithLogitsLoss, CrossEntropyLoss, MSELoss
from peft.peft_model import PeftModel
from transformers.modeling_outputs import ImageClassifierOutput, ImageClassifierOutputWithNoAttention
from transformers import ViTForImageClassification, MobileViTForImageClassification
from functools import partial

class PeftModelFixed(PeftModel):
    def forward(self, *args: Any, **kwargs: Any):
        """
        Forward pass of the model.
        """
        with self._enable_peft_forward_hooks(*args, **kwargs):
            # TODO removed this to avoid mixing
            # kwargs = {k: v for k, v in kwargs.items() if k not in self.special_peft_forward_args}
            return self.get_base_model()(*args, **kwargs)

class MobileViTForImageClassificationFixed(MobileViTForImageClassification):
    def forward(
        self,
        pixel_values: Optional[torch.Tensor] = None,
        output_hidden_states: Optional[bool] = None,
        labels: Optional[torch.Tensor] = None,
        return_dict: Optional[bool] = None,
        **kwargs # TODO added kwargs
    ) -> Union[tuple, ImageClassifierOutputWithNoAttention]:
        r"""
        labels (`torch.LongTensor` of shape `(batch_size,)`, *optional*):
            Labels for computing the image classification/regression loss. Indices should be in `[0, ...,
            config.num_labels - 1]`. If `config.num_labels == 1` a regression loss is computed (Mean-Square loss). If
            `config.num_labels > 1` a classification loss is computed (Cross-Entropy).
        """
        return_dict = return_dict if return_dict is not None else self.config.use_return_dict

        # TODO here
        outputs = self.mobilevit(pixel_values, output_hidden_states=output_hidden_states, return_dict=return_dict)

        pooled_output = outputs.pooler_output if return_dict else outputs[1]

        # TODO here
        adapter_names = kwargs["adapter_names"]
        patch_size = self.config.patch_size
        multiply = patch_size ** 2
        adapter_names_original = []
        for i in range(0, len(adapter_names), multiply):
            adapter_names_original.append(adapter_names[i])
        logits = self.classifier(self.dropout(pooled_output), adapter_names=adapter_names_original)

        loss = None
        if labels is not None:
            if self.config.problem_type is None:
                if self.num_labels == 1:
                    self.config.problem_type = "regression"
                elif self.num_labels > 1 and (labels.dtype == torch.long or labels.dtype == torch.int):
                    self.config.problem_type = "single_label_classification"
                else:
                    self.config.problem_type = "multi_label_classification"

            if self.config.problem_type == "regression":
                loss_fct = MSELoss()
                if self.num_labels == 1:
                    loss = loss_fct(logits.squeeze(), labels.squeeze())
                else:
                    loss = loss_fct(logits, labels)
            elif self.config.problem_type == "single_label_classification":
                loss_fct = CrossEntropyLoss()
                loss = loss_fct(logits.view(-1, self.num_labels), labels.view(-1))
            elif self.config.problem_type == "multi_label_classification":
                loss_fct = BCEWithLogitsLoss()
                loss = loss_fct(logits, labels)

        if not return_dict:
            output = (logits,) + outputs[2:]
            return ((loss,) + output) if loss is not None else output

        return ImageClassifierOutputWithNoAttention(
            loss=loss,
            logits=logits,
            hidden_states=outputs.hidden_states,
        )



def peftforward(self, *args, **kwargs):
    if self.disable_adapters or (self.active_adapter not in self.modules_to_save):
        return self.original_module(*args, **kwargs)

    # TODO changed to support LoRA
    adapter_names = kwargs["adapter_names"]
    kwargs = {}
    batch = args[0]
    unique_adapters = set(adapter_names)
    sub_batch_indices_list = []
    for adapter in unique_adapters:
        sub_batch_indices_list.append([index for index, item in enumerate(adapter_names) if item == adapter])

    results = [0 for i in range(len(batch))]
    for i, active_adapter in enumerate(unique_adapters):
        sub_batch = batch[sub_batch_indices_list[i]]
        output = self.modules_to_save[active_adapter](*(sub_batch,), **kwargs)
        for index, j in enumerate(sub_batch_indices_list[i]):
            results[j] = output[index]
    return torch.stack(results)

def change_forward_dynamically(model: PeftModel):
    model.classifier.forward = partial(peftforward, model.classifier)
    return model
        if not return_dict:
            output = (logits,) + outputs[1:]
            return ((loss,) + output) if loss is not None else output

        return ImageClassifierOutput(
            loss=loss,
            logits=logits,
            hidden_states=outputs.hidden_states,
            attentions=outputs.attentions,
        )


class MobileViTForImageClassificationFixed(MobileViTForImageClassification):
    def forward(
        self,
        pixel_values: Optional[torch.Tensor] = None,
        output_hidden_states: Optional[bool] = None,
        labels: Optional[torch.Tensor] = None,
        return_dict: Optional[bool] = None,
        **kwargs # TODO added kwargs
    ) -> Union[tuple, ImageClassifierOutputWithNoAttention]:
        r"""
        labels (`torch.LongTensor` of shape `(batch_size,)`, *optional*):
            Labels for computing the image classification/regression loss. Indices should be in `[0, ...,
            config.num_labels - 1]`. If `config.num_labels == 1` a regression loss is computed (Mean-Square loss). If
            `config.num_labels > 1` a classification loss is computed (Cross-Entropy).
        """
        return_dict = return_dict if return_dict is not None else self.config.use_return_dict

        # TODO here
        outputs = self.mobilevit(pixel_values, output_hidden_states=output_hidden_states, return_dict=return_dict)

        pooled_output = outputs.pooler_output if return_dict else outputs[1]

        # TODO here
        adapter_names = kwargs["adapter_names"]
        patch_size = self.config.patch_size
        multiply = patch_size ** 2
        adapter_names_original = []
        for i in range(0, len(adapter_names), multiply):
            adapter_names_original.append(adapter_names[i])
        logits = self.classifier(self.dropout(pooled_output), adapter_names=adapter_names_original)

        loss = None
        if labels is not None:
            if self.config.problem_type is None:
                if self.num_labels == 1:
                    self.config.problem_type = "regression"
                elif self.num_labels > 1 and (labels.dtype == torch.long or labels.dtype == torch.int):
                    self.config.problem_type = "single_label_classification"
                else:
                    self.config.problem_type = "multi_label_classification"

            if self.config.problem_type == "regression":
                loss_fct = MSELoss()
                if self.num_labels == 1:
                    loss = loss_fct(logits.squeeze(), labels.squeeze())
                else:
                    loss = loss_fct(logits, labels)
            elif self.config.problem_type == "single_label_classification":
                loss_fct = CrossEntropyLoss()
                loss = loss_fct(logits.view(-1, self.num_labels), labels.view(-1))
            elif self.config.problem_type == "multi_label_classification":
                loss_fct = BCEWithLogitsLoss()
                loss = loss_fct(logits, labels)

        if not return_dict:
            output = (logits,) + outputs[2:]
            return ((loss,) + output) if loss is not None else output

        return ImageClassifierOutputWithNoAttention(
            loss=loss,
            logits=logits,
            hidden_states=outputs.hidden_states,
        )

def peftforward(self, *args, **kwargs):
    if self.disable_adapters or (self.active_adapter not in self.modules_to_save):
        return self.original_module(*args, **kwargs)

    # TODO changed to support LoRA
    adapter_names = kwargs["adapter_names"]
    kwargs = {}
    batch = args[0]
    unique_adapters = set(adapter_names)
    sub_batch_indices_list = []
    for adapter in unique_adapters:
        sub_batch_indices_list.append([index for index, item in enumerate(adapter_names) if item == adapter])

    results = [0 for i in range(len(batch))]
    for i, active_adapter in enumerate(unique_adapters):
        sub_batch = batch[sub_batch_indices_list[i]]
        output = self.modules_to_save[active_adapter](*(sub_batch,), **kwargs)
        for index, j in enumerate(sub_batch_indices_list[i]):
            results[j] = output[index]
    return torch.stack(results)

def change_forward_dynamically(model: PeftModel):
    model.classifier.forward = partial(peftforward, model.classifier)
    return model
```

**Comment by BenjaminBossan** [comment]
> However, this is just a hack and I think this should work out of the box. I'm happy to investigate further when I get a chance to first solve #1960 .

Thanks a lot again for this detailed analysis, and again I would be very happy to accept a PR to fix this. Regarding the question of how to fix this: I wonder if it would be easier to change the logic inside of `_check_forward_args`. Maybe we can relax the `len(x) != len(adapter_names)` check or even (re)move it entirely if that's enough to fix the situation.

**Comment by saeid93** [comment]
No problem!
About your question, as far as I understand removing the check won't solve the problem since there will be a mismatch between input size and the number of items in the adapter_names. Therefore in the [lora_layer](https://github.com/huggingface/peft/blob/27833a2e60274aae1cfc638d2b536782e7ec3f7a/src/peft/tuners/lora/layer.py#L332) the LoRA weights are only applied proportional to the number of items in `adapter_names` (same as the input batch size) which is fine in most models. However, in MobileVit due to the unfolding operation the dimension of the inputs is different from the number of the batch size, since `adapter_names` length matches the batch the LoRA is not applied to all the inputs.

I also checked the accuracy without multi LoRA inference and it seems that my above explanation can also be validated by looking at accuracies. Removing the check does not match the accuracy of single LoRA inference.

**Comment by BenjaminBossan** [comment]
I see, I thought it would be possible to remove the check or at least make it optional. The user then needs to ensure that the correct `adapter_names` are passed so that they are lined up with the way that MobileVit unfolds the 0th dimension. The changes to `MobileViTForImageClassification` would probably still be necessary (though the mixed batch feature is intended for inference only, not sure if that simplifies things). But I'm probably missing something. Anyway, a PR to fix the situation would be welcome.

**Comment by saeid93** [comment]
Hi @BenjaminBossan ,
This issue will not going to be solved by my recent pull request. That pull requests solved https://github.com/huggingface/peft/issues/1960 . This one is a MobileViT specific problem which I was thinking to look for a potential solution next. Please let me know if you think it is no longer necessary, I just want to let you know that it is not relevant to #1990.

**Comment by saeid93** [comment]
ah, I see :) Sure, I'll have a look when I get a chance.

**Comment by saeid93** [comment]
Hi @BenjaminBossan 

I investigated this issue deeper, however, I'm a bit uncertain about the correct way to implement this. I have investigated three different approaches but they both have some problems, I have explained below, please let me know about your opinion them and any other solutions that you suggest.

### Background
As I explained above the problem is that the unfolding operation changes the dimensions of the input in the MobileViT and therefore we have to increase the `adapter_names` proportional to the patch size of the input:
```python
# -------- changing the size of the adapter_names input ----------
        if model.base_model.model.base_model_prefix == "mobilevit":
            patch_size = model.config.patch_size
            multiply = patch_size ** 2
            resized_adapters_names = []
            for item in batch["adapter_names"]:
                multiplied = [item] * multiply
                resized_adapters_names += multiplied
            batch["adapter_names"] = resized_adapters_names
        outputs = model(**batch)
```
Note that after the fixes https://github.com/huggingface/peft/pull/1990 this solution won't work out of the box since the [mobilevit part](https://github.com/huggingface/transformers/blob/5ef432e4742cc505f610f8e54ac1cd2e1dfd265e/src/transformers/models/mobilevit/modeling_mobilevit.py#L827) expects the changed format above and the [classifier part](https://github.com/huggingface/transformers/blob/5ef432e4742cc505f610f8e54ac1cd2e1dfd265e/src/transformers/models/mobilevit/modeling_mobilevit.py#L831) part expects the original length for `adapter_names` input.

### Solution 1
Trying to interfere somewhere in the code and change the `adapter_names` layer differently for ViT part and classifier part 

This solution aimed to basically do the hack I'm doing above (subclassing the mobilevit) without the subclassing and injecting the changed parts dynamically similar to [here](https://github.com/huggingface/peft/blob/4611034ff85e26e1f9647ea1f505e9e50322ff0f/src/peft/tuners/lora/model.py#L434-L438) through a prehook that can change the `adapter_names` if the model type if `mobilevit`. However, the problem I faced was to add the logic of how much the size of `adapter_names` should change we need access to the `patch_size` variable in some function like [this one](https://github.com/huggingface/peft/blob/4611034ff85e26e1f9647ea1f505e9e50322ff0f/src/peft/tuners/lora/model.py#L59) to be later added as a prehook. However, when I implemented this halfway passing down that variable the patch_size variable needed lots of changes to the existing function signatures.

### Solution 2
Rewrite how the Peft library changes the LoRA layers

In this one I aimed to rewrite the [SelfAttention layer](https://github.com/huggingface/transformers/blob/5ef432e4742cc505f610f8e54ac1cd2e1dfd265e/src/transformers/models/mobilevit/modeling_mobilevit.py#L232) of MobileViT to accommodate this change of size when LoRA is applied through [adding a dispatcher](https://github.com/huggingface/peft/blob/8d9ecbed080eed79f32b279bdec211767919bd94/src/peft/tuners/lora/model.py#L330) for MobileViT. The problem though was adding this required changing the logic of adding LoRA layers which I think can potentially break some other parts.

### Solution 3
Re-write the MobileVit with an inherited function in the Peft library (similar to the hack I did above but with some modification to accommodate fixes of #1990). However, the problem with this solution is that it sounds super hacky to have a special logic for a specific model type in Peft library.

Please let me know if you have any suggestions for a better approach or any comments about the discussed solutions. I'm happy to continue based on your suggestions.

**Comment by saeid93** [comment]
Thank you for the suggestion! I'll take a closer look when I get the chance. One quick question that comes to mind: since `adapter_names` is a list and doesn't support broadcasting, are you suggesting that everything should be converted to NumPy arrays before attempting this solution?

**Comment by BenjaminBossan** [comment]
No, I don't really mean broadcasting in the sense of numpy, hence why I wrote "broadcasting" :) What I mean is repeating the same items multiple times. Simplified code would be something like this:

```python
adapter_names = ["a", "b", "a", "c"]
x = range(12)  # 3 times the size of adapter_names
quot, remainder = divmod(len(adapter_names), len(x))  # 3, 0
if remainder != 0:
    raise ...
adapter_names = sum([[i] * quot for i in adapter_names], [])
print(adapter_names)  # ['a', 'a', 'a', 'b', 'b', 'b', 'a', 'a', 'a', 'c', 'c', 'c']
```

**Comment by saeid93** [comment]
Thank you for your clarification, I'll work on it when I get a chance.

**Comment by saeid93** [comment]
@BenjaminBossan Thank you for your comment and for proposing a solution. Having multiple classifier heads in the base model based on the LoRA requirements would indeed solve my issue. As you suggested, I could remove the base heads and keep only those needed for the LoRAs. Do you think this is something worth exploring as a feature for the main library? For example, instead of modifying the `modules_to_save` as I initially suggested, we could introduce a new concept or class, such as `dynamic_adapter_modules`, which could be specified for each LoRA and might not necessarily exist in the base model.

**Comment by BenjaminBossan** [comment]
Great, thanks. Of course I might be missing something and one of your proposals could make more sense.
